### PR TITLE
feat: Publish stream via stdin & Unix socket

### PR DESCRIPTION
**Pull request by dwirya**
Summary: 
- Requires `https://github.com/livekit/server-sdk-go/pull/64` to be merged and version bumped to `v0.10.1`
- Publishing arguments are more specific now for consistency: `publish-file`, `publish-demo` and `publish-stream`.

Testing:
- Using file pipe: 
  `cat matrix-trailer.h264 |
        livekit-cli join-room --room {room} --identity {identity}
        --publish-stream video/h264 --fps 25
        --api-key {api_key}
        --api-secret {api_secret}`
- Using ffmpeg:
  `ffmpeg -i matrix-trailer.h264 -f h264 - |
        livekit-cli join-room --room {room} --identity {identity}
        --publish-stream video/h264 --fps 25
        --api-key {api_key}
        --api-secret {api_secret}
`

### CODE REVIEW

**Comment by dwirya** [comment]
File pipe (`cat`) works well and fine, but using GStreamer & FFMPEG resulted in glitchy videos. Not sure why, hoping that someone can point me in the right direction

**Comment by dwirya** [comment]
The following FFMPEG command works: `ffmpeg -i matrix-trailer.h264 -bf 0 -max_delay 0 -f h264 - | livekit-cli join-room --room {room} --identity {identity} \
--publish-stream video/h264 --fps 23.98 \
--api-key {api_key} --api-secret {api_secret}`. 

`-bf 0` is needed to disable B-frames, while `-max_delay 0` is to disable buffering and writing the packets in chunks.

**Comment by dwirya** [comment]
The updated CLI will have 4 modes: `publish-demo`, `publish-file`, `publish-stdin` and `publish-socket`. `publish-stdin` is proposed to be available as it's more convenient than `publish-socket`, where user needs to set up Unix socket sender (see README)

**Comment by davidzhao** [comment]
@dwirya sorry for the late review here. been thinking about this for a bit. I think offering unix socket would be a great way to go, but I'm wondering if we can do so without having to change the existing CLI parameters. A couple of requests:

1. can we keep the --publish syntax, and detect if it's a unix socket that's passed in or not. For example, if you passed in /path/to/file, we'd treat it as a regular file, but if you did unix:///path/to/socket, then we would read it like a unix socket.
2. can we give an example that's creating separate streams with ffmpeg writing to unix socket, then reading with livekit? i.e. would something [documented here](https://lists.ffmpeg.org/pipermail/ffmpeg-user/2017-May/036134.html) work?

**Comment by dwirya** [comment]
@davidzhao that's alright, I've been busy too. Regarding the requests:
1. I think it's a good idea, so I'm getting this correctly, eventually we'll only have 2 modes: `--publish` and `--publish-demo`
2. It's definitely possible. I actually wanted to provide an example for `cat` and `ffmpeg`, but I couldn't get the audio working with `ffmpeg`. I'm suspecting it's due to DTX enabled (audio was very short compared to video based on debug logs) but I didn't know how to disable it. I'll do a bit more investigation and update you once it's ready

**Comment by davidzhao** [comment]
> 1. I think it's a good idea, so I'm getting this correctly, eventually we'll only have 2 modes: `--publish` and `--publish-demo`

Yup, thinking we could keep the original params to keep it consistent.

> 2. It's definitely possible. I actually wanted to provide an example for `cat` and `ffmpeg`, but I couldn't get the audio working with `ffmpeg`. I'm suspecting it's due to DTX enabled (audio was very short compared to video based on debug logs) but I didn't know how to disable it. I'll do a bit more investigation and update you once it's ready

What does your command line look like for ffmpeg? you don't have to encode opus with dtx, I think that's fine. For video playback, the --fps flag would need to be passed in.

**Comment by dwirya** [comment]
@davidzhao I've implemented Feedback 1, so `--publish` takes in a filename (e.g. `--publish matrix-trailer.h264`) or a Unix socket address (e.g. `--publish unix:/tmp/ffmpeg-h264.sock`). 

As for Feedback 2, I haven't had much luck publishing just Opus via socket. The command used was `ffmpeg -nostdin -i matrix-trailer.ogg -max_delay 0 -f opus -listen 1 unix:/tmp/ffmpeg-opus.sock`. 

I tried using video from socket + audio from file, so something like `--publish unix:/tmp/ffmpeg-h264.sock --publish matrix-trailer.ogg --fps 23.98` and it works. I'd appreciate any pointers 🙏

**Comment by davidzhao** [comment]
I tried it today and it's working really well! I got audio to work with this command line:

```
ffmpeg -i matrix-trailer2.ogg -c:a libopus -page_duration 20000 -nostdin -listen 1 -f opus unix:///tmp/ffmpeg-opus.sock
```

it could take from any input, I just had a ogg available. Will try it some more tomorrow with a full example transcoding both video/audio from a mp4.

**Comment by dwirya** [comment]
@davidzhao thanks for following up on this 🙏! I’ll play around with `gst-launch` as well for completeness.

**Review by davidzhao — APPROVED** [review]
lg Dennis! awesome work

**Comment by davidzhao** [comment]
looks like we have a couple of conflicts to resolve here.

**Comment by dwirya** [comment]
I think there were a few merges after my PR. Not sure what the best steps are, maybe I should pull from the latest and put back my code. Do you think that’ll solve the problem ? 

FYI I don’t have access to WiFi until Monday, so apologies for the delay in responding.

**Comment by davidzhao** [comment]
no problem, I can take a look too.

**Comment by dwirya** [comment]
Thanks David!

**Comment by bsirang** [comment]
@davidzhao / @dwirya I was hoping to use the functionality described here to pipe a gstreamer pipeline over to livekit-cli. As I read the discussion here about the `--publish` flag, I can't tell if the stdin piping feature described originally still exists.

So I have a few questions:
1) Can we pipe data to livekit-cli via stdin?
2) Do you have an example of piping h264 via gstreamer into livekit-cli? I was thinking of using `filesink location=/dev/stdout` and piping that to the cli
3) Do you plan on supporting shm via gstreamer's shmsink?

**Comment by dwirya** [comment]
Hi @bsirang ,

Sorry for the late reply. I've been busy at work so haven't had much time to contribute. I've had someone ask me regarding gstreamer integration, but the in-built socket plugins, `shmsink` and `shmsrc` don't seem to work well. I tried streaming `videotestsrc` to a Unix socket using `shmsink`, and use `shmsrc` to receive from socket and pipe it to `ffplay`. `ffplay` will only play it if the stream is raw video; using encoded stream such as `h264` and `vp8` gave me an error on `ffplay`. 

The `stdin` feature worked well for gstreamer integration, but it required a new API (something like `publish-stdin`) to tell the CLI which codec reader to use. I believe my discussion with @davidzhao was to keep the API simple, and Unix socket seems to suffice at the time. We can bring it back if required.

In summary,
1. Currently no, but it's a pretty simple change. I can probably do it by tomorrow (at @davidzhao's discretion to merge it)
2. I did try to pipe `gstreamer`'s H264 stream to the CLI via `stdin`, so I can give an example in the README
3. Probably not possible at this point, because I've seen some problems on my end with `shmsink` when publishing encoded stream such as H264. My version is 1.18
